In [2]:
import cx_Oracle
import sys
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import argparse
import os
import IPython
from IPython.display import Image, display
import time

In [3]:
import pascal_utils as pas

Let's see all the available schemas, and the tables that are available in each. 

```javascript
SCHEMAS = ['CMS_HGC_CORE_ATTRIBUTE',
           'CMS_HGC_CORE_COND', 
           'CMS_HGC_HGCAL_COND', 
           'CMS_HGC_CORE_CONSTRUCT', 
           'CMS_HGC_CORE_IOV_MGMNT', 
           'CMS_HGC_CORE_MANAGEMNT']
```
Since all start with `CMS_HGC`, let's remove that from all, so
```javascript
SCHEMAS = [
            'CORE_ATTRIBUTE',
           (parts and data types)
           'CORE_COND', 
            (user-generated data, like tests)
            'HGCAL_COND', 
            (user-generated data spicific to HGCAL, like tests)
           'CORE_CONSTRUCT', 
           (parts)
               'CORE_IOV_MGMNT', 
                   (interval of validity of user-generated data)
           'CORE_MANAGEMNT'
            (locations, institutions, etc.)
    ]
```

In [16]:
pas.execute_query(
"""
SELECT table_name, num_rows from all_tables WHERE OWNER='CMS_HGC_HGCAL_COND'
ORDER BY num_rows DESC
""")


COLUMN NAMES:
 ['TABLE_NAME', 'NUM_ROWS']
\FINISHED 'execute_query' in 3.3594 SECS


TABLE_NAME  NUM_ROWS
0         HGC_SENSOR_DEFECT_CHKS       NaN
1               HGC_PQC_DIODE_IV       NaN
2               HGC_PQC_DIODE_CV       NaN
3                    HGC_PQC_FET       NaN
4      HGCROC_DACB_CONVEYOR_TEST       NaN
5          HGCROC_INJECTION_SCAN       NaN
6            HGCROC_PEDESTAL_RUN       NaN
7              HGCROC_PHASE_SCAN       NaN
8           HGCROC_RAM_RETENTION       NaN
9           HGCROC_SAMPLING_SCAN       NaN
10        HGCROC_TDC_TOA_THRSHLD       NaN
11        HGCROC_TDC_TOT_THRSHLD       NaN
12      HGC_PRTO_MOD_ASMBLY_COND       NaN
13                   HGCROC_VREF       NaN
14          HGC_PQC_VAN_DER_PAUW       NaN
15                 HGC_PQC_SUMRY       NaN
16       HGC_PQC_OXIDE_BREAKDOWN       NaN
17                   HGC_PQC_MOS       NaN
18             HGC_PQC_LINEWIDTH       NaN
19           HGC_MOD_ASMBLY_COND       NaN
20         HGC_MOD_WIREBOND_TEST       NaN
21                   HGC_PQC_GCD       NaN
22            HGC_CERN_SENSOR_IV  300716.0
23            HGC_CERN_SENSOR_CV   88147.0
24                 HGC_SENSOR_IV   22334.0
25                TEST_SENSOR_IV    3400.0
26                TEST_SENSOR_CV    2992.0
27                 HGC_SENSOR_CV     618.0
28      UCSB_CHAN_NOISE_SUMRY_HG     264.0
29      UCSB_CHAN_NOISE_SUMRY_LG     256.0
30      HGC_SIXINCH_MOD_CHAN_MAP     128.0
31      HGC_CERN_SENSOR_CV_SUMRY      85.0
32      HGC_CERN_SENSOR_IV_SUMRY      76.0
33           SI_MODULE_BASEPLATE      70.0
34          FLATNS_PCB_ROCS_DATA      43.0
35        HGC_BARE_PCB_ELEC_DATA      38.0
36             HGC_BARE_PCB_DATA      38.0
37  HGC_SENSOR_IRRADIATION_SUMRY      29.0
38            FLATNS_SENSOR_DATA       7.0
39                HGC_MOD_ASMBLY       6.0
40           HGC_PRTO_MOD_ASMBLY       6.0
41                   HGCROC_TEST       3.0
42            HGC_NOISE_MAP_BLOB       2.0
43        TEST_SENSOR_CONDITIONS       1.0

See the registered wafer that has serial number "100113":

In [17]:
pas.execute_query("select * from CMS_HGC_CORE_CONSTRUCT.PARTS where SERIAL_NUMBER='100113'")


COLUMN NAMES:
 ['PART_ID', 'KIND_OF_PART_ID', 'LOCATION_ID', 'MANUFACTURER_ID', 'IS_RECORD_DELETED', 'RECORD_INSERTION_TIME', 'RECORD_INSERTION_USER', 'BARCODE', 'SERIAL_NUMBER', 'VERSION', 'NAME_LABEL', 'INSTALLED_DATE', 'REMOVED_DATE', 'INSTALLED_BY_USER', 'REMOVED_BY_USER', 'EXTENSION_TABLE_NAME', 'COMMENT_DESCRIPTION', 'RECORD_LASTUPDATE_TIME', 'RECORD_LASTUPDATE_USER', 'PRODUCTION_DATE', 'BATCH_NUMBER']
\FINISHED 'execute_query' in 2.5503 SECS


PART_ID  KIND_OF_PART_ID  LOCATION_ID  MANUFACTURER_ID IS_RECORD_DELETED  \
0    25990            11080         5780             1000                 F   

  RECORD_INSERTION_TIME          RECORD_INSERTION_USER BARCODE SERIAL_NUMBER  \
0   2022-08-10 21:08:33  Alexander Joseph Wade (awade)    None        100113   

  VERSION  ... INSTALLED_DATE REMOVED_DATE INSTALLED_BY_USER REMOVED_BY_USER  \
0    None  ...           None         None              None            None   

  EXTENSION_TABLE_NAME       COMMENT_DESCRIPTION RECORD_LASTUPDATE_TIME  \
0                 None  300um LD Si Wafer 100113                   None   

          RECORD_LASTUPDATE_USER PRODUCTION_DATE BATCH_NUMBER  
0  Alexander Joseph Wade (awade)            None         None  

[1 rows x 21 columns]

We see that it was uploaded by (`RECORD_LASTUPDATE_USER`) Alex Wade, and where it was tested (`LOCATION_ID`) is 5780.

Now, what is this location in words?

In [4]:
pas.execute_query(
"""
SELECT table_name, num_rows from all_tables WHERE OWNER='CMS_HGC_CORE_MANAGEMNT'
ORDER BY num_rows DESC
""")


COLUMN NAMES:
 ['TABLE_NAME', 'NUM_ROWS']

FINISHED 'execute_query' in 3.2142 SECS


TABLE_NAME  NUM_ROWS
0                     USERS       NaN
1             LOCATIONS_HST       NaN
2  CONDITIONS_DATA_AUDITLOG    3657.0
3                 LOCATIONS      40.0
4              INSTITUTIONS      34.0
5          INSTITUTIONS_HST      18.0

In [5]:
pas.execute_query(
"""
SELECT * FROM CMS_HGC_CORE_MANAGEMNT.INSTITUTIONS
""")


COLUMN NAMES:
 ['INSTITUTION_ID', 'IS_RECORD_DELETED', 'INSTITUTE_CODE', 'RECORD_INSERTION_TIME', 'RECORD_INSERTION_USER', 'RECORD_LASTUPDATE_TIME', 'RECORD_LASTUPDATE_USER', 'NAME', 'COMMENT_DESCRIPTION', 'TOWN', 'COUNTRY']

FINISHED 'execute_query' in 2.8045 SECS


INSTITUTION_ID IS_RECORD_DELETED  INSTITUTE_CODE RECORD_INSERTION_TIME  \
0             3140                 T             0.0   2021-09-30 14:44:55   
1             3960                 F             1.0   2021-10-12 02:26:37   
2             5120                 F             0.0   2022-08-24 18:05:54   
3             5121                 F             0.0   2022-08-24 18:05:57   
4             5122                 F             0.0   2022-08-24 18:06:06   
5             1500                 F             0.0   2017-12-07 23:03:43   
6             1501                 F             0.0   2017-12-07 23:03:43   
7             1502                 F             0.0   2017-12-07 23:03:44   
8             1503                 F             0.0   2017-12-07 23:03:44   
9             1504                 F             0.0   2017-12-07 23:03:45   
10            2740                 F             NaN   2021-01-14 06:33:06   
11            2760                 T             NaN   2021-01-14 06:48:45   
12            4360                 F             1.0   2021-10-12 02:29:37   
13            5920                 F             0.0   2022-11-20 21:49:26   
14            6740                 F             0.0   2023-04-05 17:58:39   
15            6760                 F             0.0   2023-04-05 18:09:14   
16            7520                 F             0.0   2023-04-19 15:13:11   
17            7521                 F             0.0   2023-04-19 15:13:11   
18            1920                 F             0.0   2020-03-21 01:59:42   
19            3540                 F             NaN   2021-10-01 14:42:55   
20            4340                 F             1.0   2021-10-12 02:29:12   
21            5940                 F             0.0   2022-11-21 00:15:25   
22            7120                 F             0.0   2023-04-05 18:11:27   
23            7540                 F             0.0   2023-04-19 15:28:30   
24            1520                 F             0.0   2018-03-26 00:20:27   
25            4380                 F             1.0   2021-10-12 02:30:37   
26            4720                 F             0.0   2022-08-10 21:08:32   
27            1000                 F             NaN   2017-09-14 06:09:41   
28            1420                 F             1.0   2017-12-06 05:53:56   
29            1440                 F             NaN   2017-12-06 05:58:42   
30            1460                 F             0.0   2017-12-06 05:58:47   
31            1480                 F             NaN   2017-12-06 06:22:08   
32            3940                 F             1.0   2021-10-12 02:22:59   
33            5520                 F             0.0   2022-11-10 10:05:23   
34            7920                 F             0.0   2023-06-13 21:42:09   

            RECORD_INSERTION_USER RECORD_LASTUPDATE_TIME  \
0          CMS_HGC_CORE_MANAGEMNT    2022-02-15 23:09:43   
1            Alethea Butler-Nalin    2021-10-12 03:29:39   
2   Alexander Joseph Wade (awade)                    NaT   
3   Alexander Joseph Wade (awade)                    NaT   
4   Alexander Joseph Wade (awade)                    NaT   
5    CMS_HGC_PRTTYPE_HGCAL_WRITER                    NaT   
6    CMS_HGC_PRTTYPE_HGCAL_WRITER                    NaT   
7    CMS_HGC_PRTTYPE_HGCAL_WRITER                    NaT   
8    CMS_HGC_PRTTYPE_HGCAL_WRITER                    NaT   
9    CMS_HGC_PRTTYPE_HGCAL_WRITER                    NaT   
10                          Umesh    2021-01-14 06:44:26   
11                          Umesh    2021-01-14 06:50:10   
12           Alethea Butler-Nalin    2021-10-12 03:29:53   
13                          Umesh                    NaT   
14  Alexander Joseph Wade (awade)                    NaT   
15  Alexander Joseph Wade (awade)                    NaT   
16  Alexander Joseph Wade (awade)                    NaT   
17  Alexander Joseph Wade (awade)                    NaT   
18   CMS_HGC_PRTTYPE_HGCAL_WRITER                    Na

So, the question 
"where was sensor with serial number 100113 tested?"
translates to the query


In [7]:
pas.execute_query("""
select * from CMS_HGC_CORE_CONSTRUCT.PARTS,
CMS_HGC_CORE_MANAGEMNT.INSTITUTIONS.LOCATION_ID
where SERIAL_NUMBER='100113' 
INNER JOIN CMS_HGC_CORE_CONSTRUCT.PARTS.RECORD_INSERTION_USER on 
CMS_HGC_CORE_CONSTRUCT.PARTS.RECORD_INSERTION_USER = FROM CMS_HGC_CORE_MANAGEMNT.INSTITUTIONS
""")

DatabaseError: ORA-00933: SQL command not properly ended